src/Features/Core/Portable/ExtractMethod/ExtractMethodResult.cs

```diff
@@ -27,69 +22,32 @@ internal abstract class ExtractMethodResult
         /// </summary>
         public ImmutableArray<string> Reasons { get; }
 
-        /// <summary>
-        /// The transformed document that was produced as a result of the extract method operation.
-        /// </summary>
-        public Document? DocumentWithoutFinalFormatting { get; }
-
-        /// <summary>
-        /// Formatting rules to apply to <see cref="DocumentWithoutFinalFormatting"/> to obtain the final formatted
-        /// document.
-        /// </summary>
-        public ImmutableArray<AbstractFormattingRule> FormattingRules { get; }
-
-        /// <summary>
-        /// The name token for the invocation node that replaces the extracted code.
-        /// </summary>
-        public SyntaxToken? InvocationNameToken { get; }
+        private readonly AsyncLazy<(Document document, SyntaxToken? invocationNameToken)>? _lazyData;
 
         internal ExtractMethodResult(
             OperationStatusFlag status,
             ImmutableArray<string> reasons,
-            Document? documentWithoutFinalFormatting,
-            ImmutableArray<AbstractFormattingRule> formattingRules,
-            SyntaxToken? invocationNameToken)
+            Func<CancellationToken, Task<(Document document, SyntaxToken? invocationNameToken)>>? getDocumentAsync)
         {
-            Status = status;
-
             Succeeded = status.Succeeded();
 
             Reasons = reasons.NullToEmpty();
 
-            DocumentWithoutFinalFormatting = documentWithoutFinalFormatting;
-            FormattingRules = formattingRules;
-            InvocationNameToken = invocationNameToken;
+            if (getDocumentAsync != null)
+                _lazyData = AsyncLazy.Create(getDocumentAsync);
         }
 
-        /// <summary>
-        /// internal status of result. more fine grained reason why it is failed. 
-        /// </summary>
-        internal OperationStatusFlag Status { get; }
+        public static ExtractMethodResult Fail(OperationStatus status)
+            => new(status.Flag, status.Reasons, getDocumentAsync: null);
 
-        public async Task<(Document document, SyntaxToken? invocationNameToken)> GetFormattedDocumentAsync(CodeCleanupOptions cleanupOptions, CancellationToken cancellationToken)
+        public static ExtractMethodResult Success(
+            OperationStatus status,
+            Func<CancellationToken, Task<(Document document, SyntaxToken? invocationNameToken)>> getDocumentAsync)
         {
-            if (DocumentWithoutFinalFormatting is null)
-                throw new InvalidOperationException();
-
-            var annotation = new SyntaxAnnotation();
-
-            var root = await DocumentWithoutFinalFormatting.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
all this logic moved inside MethodExtractor as the last thing it does in the lazy async callback it passes to the result object.